(Naive) Early Stopping

ManuelLerchner opened the pull request:
# (Naive) Early Stopping

## Problem

The Autotuning process can be very costly at times, as many "bad" configurations can end up being evaluated.
Many of those bad configurations are orders of magnitude slower than the optimal one (see [Tobias Humig Project Report](https://mediatum.ub.tum.de/doc/1730215/qwnftyj83fo5se9xzl0lokq4a.pdf), [Manuel Lerchner Thesis](https://mediatum.ub.tum.de/doc/1752754/j9j37ciousd48fg4av11hu4s1.pdf)), which can drastically increase the total runtime of the simulation.

## Change

This PR introduces an improvement to this problem by preventing such bad configurations from being resampled multiple times during the tuning phases.
It works by detecting `long-running` samples and simply skipping to the next configuration in the configuration queue. (See next graphic).

![image](https://github.com/user-attachments/assets/aeffbf5b-6f0f-42e8-9670-dc9d3d65214a)

The current detection of `long-running` samples is quite naive. It works by comparing the slowdown ($$slowdown = \frac{time_{curr\ sample}}{time_{best\ sample\ so\ far}}$$) to a user-defined threshold called $allowedSlowdownFactor$.

## Benchmark: Exploding Liquid + PredictiveTuning (CoolMUC2)

Early tests suggest that a good threshold is around $$allowedSlowdownFactor=5$$ (see next chart), although I suspect the value is not universal and might change depending on the hardware/scenario/tuning-strategy (More Benchmarks needed).

For the Exploding Liquid benchmark with PredictiveTuning, the total simulation time can be reduced from 28.62s to 23.23s at $$allowedSlowdownFactor ≈ 5$$, corresponding to a 18.9% speedup in total simulation time compared to not using Early Stopping. The time spent in tuning phases is reduced by ≈60% in this case.

![total_time_average_full_scale](https://github.com/user-attachments/assets/0eaf58f3-a133-474c-ba11-0ad6959741dd)

# How Has This Been Tested?

+ Benchmarks

SamNewcome reviewed (changes requested):
Generally very nice. A couple of minor points:

- I'm concerned that the parameter name is too interchangeable with the relative blacklist range used with the Slow Config filter strategy, especially if you don't look too closely at the doxygen description. I would try to make it clearer that these are separate. One option could be to call this the early stopping factor (or something like that). Alternatively, you could rename the slow config filter's parameter.
- I'm concerned that this factor could be too exclusionary of the Verlet-based methods with expensive first iterations. This probably doesn't show in the current results because Verlet-based methods appear to have implementational issues, so they are (in general) unlikely to be optimal. I would ideally want the comparison to be made on the weighted average. For the sake of simplicity, I would suggest simply requiring two iterations and then using the weighted average. This could, however, potentially be improved. I would also be okay with leaving this for this PR, but adding this as an issue to be addressed in the future.

manishmishra6016 commented on src/autopas/tuning/AutoTuner.cpp:
```diff
@@ -444,4 +462,22 @@ bool AutoTuner::inTuningPhase() const {
+void AutoTuner::evaluateConfigurationPerformance(long performance) {
+  _fastestIteratePairwiseTime = std::min(_fastestIteratePairwiseTime, performance);
+
+  double slowdownFactor = static_cast<double>(performance) / static_cast<double>(_fastestIteratePairwiseTime);
```
I have the same concern as Sam here. Because you compare only the first iteration, you bias towards the configuration that have low `rebuildIteration` time. These might have overall larger time due to a more expensive `nonRebuildIteration` time. Based on current implementations, the current approach biases towards `LinkedCells` over VL-based containers.

manishmishra6016 reviewed (commented):
I was trying to run the simulations with different `max-allowed-slowdown-factor` but I end up with the segmentation fault for most values (works fine when the factor is ~5 or higher, but fails for 4 or lower in case of fallingDrop example). I didn't look into it in any detail, but did you encounter this as well? Or am I doing something wrong?

ManuelLerchner commented:
> I was trying to run the simulations with different `max-allowed-slowdown-factor` but I end up with the segmentation fault for most values (works fine when the factor is ~5 or higher, but fails for 4 or lower in case of fallingDrop example). I didn't look into it in any detail, but did you encounter this as well? Or am I doing something wrong?

I also encountered this occasionally for certain values of `max-allowed-slowdown-factor`. 
However, I'm not sure if this is a direct result of my changes. Sam mentioned something about Verletlists being broken, so maybe it is caused by that.

Referring to our conversation on Elements, some of the crashes seem to be caused by `ClusterList.getClusterThreadPartition()` in `VCLC01BalancedTraversal.h`, as the function somehow returns invalid references, in some executions. However, I did not yet look deeper into why this is the case.

SamNewcome commented:
> Referring to our conversation on Elements, some of the crashes seem to be caused by `ClusterList.getClusterThreadPartition()` in `VCLC01BalancedTraversal.h`, as the function somehow returns invalid references in some executions. However, I did not yet look deeper into why this is the case.

I have also independently had issues with `VCLC01BalancedTraversal.h`, but didn't get a chance to look into it.

ManuelLerchner commented:
I tried a new way of calculating the slowdown factor. Now it works with `estimateRuntimeFromSamples()` in combination with `_evidenceCollection`. 
I changed it to require at least two samples, before considering an Early-Stop, and `estimateRuntimeFromSamples()` does the `weighted average` internally.

Since this method is less affected by noise, you can get away with way lower `allowed-slowdown-factors`, and somehow the "U-Shape is gone".

And I fixed the `segmentation faults`.

![total_time_average](https://github.com/user-attachments/assets/9c5eb377-cce9-4314-a1a3-78579933d207)

manishmishra6016 commented on src/autopas/tuning/AutoTuner.cpp:
```diff
@@ -463,21 +459,33 @@ const EvidenceCollection &AutoTuner::getEvidenceCollection() const { return _evi
     _earlyStoppingOfResampling = true;
+
+    // Pretend that we are in the last iteration of this sample. This is required for the periodic rebuilding
+    _iterationBaseline += (_maxSamples - (_iterationBaseline % _maxSamples)) - 1;
```
Maybe explain the formula in the comment a bit (it works fine for what I tried). But somethings are not clear, for eg. shouldn't `_iterationBaseline % _maxSamples` always be just `_iterationBaseline`?

ManuelLerchner replied:
`_iterationBaseline` is not required to be less than `_maxSamples`. It starts at 0 at the start of the tuning phase and gets incremented every iteration.

`_iterationBaseline` is primarily used in:
```cpp
bool AutoTuner::willRebuildNeighborLists() const {
  // What is the rebuild rhythm?
  const auto iterationsPerRebuild = this->inTuningPhase() ? _maxSamples : _rebuildFrequency;
  // _iterationBaseLine + 1 since we want to look ahead to the next iteration
  const auto iterationBaselineNextStep = _forceRetune ? _iterationBaseline : _iterationBaseline + 1;
  return (iterationBaselineNextStep % iterationsPerRebuild) == 0;
}
```

to trigger a rebuild whenever a new configuration is evaluated.

With Early-Stopping we want to force a rebuild in the next iteration, as a new configuration will be evaluated.

The formula simply adds the number of samples that we skipped to `_iterationBaseline`, to pretend we are at the end of the sampling phase. I will add this in the comments.

manishmishra6016 commented on src/autopas/tuning/AutoTuner.cpp:
```diff
@@ -463,21 +459,33 @@ const EvidenceCollection &AutoTuner::getEvidenceCollection() const { return _evi
-  double slowdownFactor = static_cast<double>(performance) / static_cast<double>(_fastestIteratePairwiseTime);
+void AutoTuner::checkEarlyStoppingCondition() {
+  if (_samplesNotRebuildingNeighborLists.empty()) {
+    // Wait until a sample without expensive rebuilding occurred to make it fairer for Verlet-based methods.
+    // This should generally happen in the second sample
```
```suggestion
    // Wait until a sample without the expensive rebuilding has occurred to compare the configurations based on estimated runtime for the non-tuning phase.
    // This should generally happen in the second sample.
```
Maybe better?

SamNewcome commented on examples/md-flexible/src/configuration/MDFlexConfig.h:
```diff
@@ -368,12 +368,13 @@ class MDFlexConfig {
+      std::numeric_limits<double>::infinity(), "early-stopping-factor", false,
+      "EarlyStoppingFactor for the auto-tuner. A configuration seeming to perform worse than the "
+      "previously best configuration "
+      "by this factor will not be sampled again"};
```
```suggestion
      "by this factor will not be sampled again during that tuning phase."};
```

SamNewcome commented on src/autopas/AutoPasDecl.h:
```diff
@@ -687,14 +687,11 @@ class AutoPas {
-   * tuning phase.
-   * @param maxAllowedSlowdownFactor
+   * Set the earlyStoppingFactor for the auto tuner. If a configuration seems to be slower than the optimum
+   * configuration found so far by more than this factor, it will not be sampled again
```
```suggestion
   * configuration found so far by more than this factor, it will not be sampled again during that tuning phase.
```

SamNewcome commented on src/autopas/tuning/AutoTuner.h:
```diff
@@ -229,12 +229,12 @@ class AutoTuner {
-   * Keeps track of the fastest runtime so far and triggers early stopping of the sampling phase if the current runtime
-   * is worse than the fastest runtime by more than the maximum allowed slowdown factor.
-   * @param performance The runtime of the current configuration.
+   * Checks whether the current configuration seems to perform so poorly that it shouldn't be resampled further
```
```suggestion
   * Checks whether the current configuration performs so poorly that it shouldn't be resampled further within this tuning phase.
```

SamNewcome commented on src/autopas/tuning/AutoTuner.h:
```diff
@@ -229,12 +229,12 @@ class AutoTuner {
-   * is worse than the fastest runtime by more than the maximum allowed slowdown factor.
-   * @param performance The runtime of the current configuration.
+   * Checks whether the current configuration seems to perform so poorly that it shouldn't be resampled further
+   * If the current configuration seems worse than the previosuly best configuration by more than the
```
```suggestion
   * If the currently sampled configuration is worse than the current best configuration by more than the
```

SamNewcome commented on src/autopas/tuning/AutoTuner.h:
```diff
@@ -229,12 +229,12 @@ class AutoTuner {
-   * @param performance The runtime of the current configuration.
+   * Checks whether the current configuration seems to perform so poorly that it shouldn't be resampled further
+   * If the current configuration seems worse than the previosuly best configuration by more than the
+   * earlyStoppingFactor factor, it will not be sampled again. Uses the _estimateRuntimeFromSamples() function to
```
```suggestion
   * earlyStoppingFactor factor, it will not be sampled again this tuning phase. Uses the _estimateRuntimeFromSamples() function to
```

SamNewcome commented on src/autopas/tuning/AutoTuner.h:
```diff
@@ -325,15 +325,10 @@ class AutoTuner {
-   * Maximum allowed slowdown factor for the auto-tuner. A configuration performing worse to the currently fastest
-   * iteratePairwise run by more than this factor is not considered for the next tuning phase.
+   * EarlyStoppingFactor for the auto-tuner. A configuration performing worse than the currently best configuration
+   * by more than this factor will not be sampled again
```
```suggestion
   * by more than this factor will not be sampled again this tuning phase.
```

SamNewcome reviewed (changes requested):
This is not explicitly part of your PR, but the documentation for `_iterationBaseline` isn't sufficient, which meant I was confused by some of your changes. Could you add to the documentation for this:
"Used only for triggering rebuilds when configurations switch during tuning phases, which occurs when _iterationBaseline % _maxSamples == 0. _iterationBaseline may therefore be modified to "skip" iterations e.g. when early stopping is used."
(Technically, in the master, it is also used to trigger rebuilds during the non-tuning phase, but this trigger is redundant and will be removed in #991 so I don't see the point of adding that the the doxygen just for Manish to remove it)

ManuelLerchner commented:
> One question: the total time values for previous benchmark and current one seem to be different (current one being ~12-25 while earlier it was 23-26, so are we looking at different problems? This might affect the shape too?

It is still the same benchmark (Exploding Liquid + Predictive Tuning).
+ I've just looked at the logs again, and the old plots were using 14-threads for the simulations, while the new one just uses 1. So maybe some overhead got removed.
+ And apparently they shut down CoolMUC2 in December and since then the new `serial-cluster` seems to be based on CoolMUC4 ([See here](https://doku.lrz.de/job-processing-on-the-linux-cluster-10745970.html)). Maybe this also explains some of the performance gains.

I've uploaded the raw data to my Repository, if you want to compare the logs and the used .yaml files ([AutoPas-AutoTuning-Seminar-Paper](https://github.com/ManuelLerchner/AutoPas-AutoTuning-Seminar-Paper/tree/main/data/explodingLiquid/cluster/predictiveTuningEvidenceBased)).

ManuelLerchner commented:
Maybe rewriting the `AutoTuner::willRebuildNeighborLists()` function as suggested in #783 may be appropriate now. The function is getting harder and harder to maintain, and was the source of some very annoying bugs during the development of this PR.

SamNewcome reviewed (approved):
Looks good!
